### ISSUE TITLE
Plotting Ray Tune results documentation example not functional?

### ISSUE THREAD

**Issue by JDups** [ultralytics/ultralytics#7144]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Using Ultralytics YOLOv8.0.229

I've been trying to follow the Ray Tune hyperparameter tuning process outlined in [the documentation](https://docs.ultralytics.com/integrations/ray-tune/#processing-ray-tune-results).

Specifically when it comes to plotting the results I have noticed several issues.

Here is the code I used to run my test tuning trials: 
```
result_grid = model.tune(data=f"{dataset.location}/data.yaml",
                         gpu_per_trial=1,
                         space={"lr0": tune.uniform(1e-5, 1e-1)},
                         optimizer="AdamW",
                         epochs=3,
                         grace_period=2,
                         iterations=3,
                         use_ray=True)
```

Here is the plot I get from the [snippet in the documentation](https://docs.ultralytics.com/integrations/ray-tune/#plotting-the-entire-history-of-reported-metrics-for-a-trial). I did, however, change it in the following ways: I changed the metric to be plotted to "metrics/precision(B)". In the docs there is an 'i' variable that appears without having been initialized. I assume there is a mistake and the snippet should be `for i, result in enumerate(result_grid):`?.
```
for i, result in enumerate(result_grid):
    plt.plot(result.metrics_dataframe["training_iteration"], result.metrics_dataframe["metrics/precision(B)"], label=f"Trial {i}")

plt.xlabel('Training Iterations')
plt.ylabel('metrics/precision(B)')
plt.legend()
plt.show()
```
![Untitled](https://github.com/ultralytics/ultralytics/assets/116138053/b20f283c-7a80-46af-8926-86247e150825)

The issue is that no curve gets plotted at all.

Looking at the [Ray Tune documentation](https://docs.ray.io/en/latest/tune/key-concepts.html) I am confused at what ultralytics is doing here.

The `iterations=3` seems to set the number of trials and not the number of iterations. Looking at the Ray Tune docs each trial seems to have each epoch as a n iteration, here however the results_grid has only a single iteration for each hyperparameter trial which would indeed not make it very plotable.

Any help or advice would be greatly appreciated.
Thank you

**Comment by glenn-jocher** [comment]
@JDups hello,

Thank you for reaching out and for your thorough research on this matter. It seems that you've encountered an issue with the plotting functionality in the Ray Tune integration with YOLOv8.

From your description, the absence of plotted curves might stem from incorrect or missing data points coming from `result.metrics_dataframe`. This could be a result of either an error in the way the data is being logged or retrieved, or if indeed the `iterations` parameter is not functioning as expected, thereby not generating enough data to plot.

Regarding the `iterations` parameter, it should indeed define the number of search iterations for hyperparameter tuning, rather than the number of training trials. Each iteration in Ray Tune is typically an evaluation of a hyperparameter combination, which may involve one or more training epochs depending on the configuration.

It's plausible that a single iteration per trial isn't sufficient to create a meaningful plot, especially since the `grace_period` could also be impacting the available data if trials are being stopped early.

To address the issue:
- Please ensure your `result_grid` contains the expected number of iterations and that the `metrics/precision(B)` field is populated with data. Review the shape and contents of `result.metrics_dataframe` for each trial in `result_grid`.
- Confirm that the tuning process is running for the intended number of epochs and that early stopping conditions are correctly set up so as to not prematurely terminate trials.

We'll look into clarifying the documentation and code snippet regarding the plotting of Ray Tune results. Meanwhile, examining your `result_grid` and ensuring a proper number of iterations should help in progressing towards a resolution.

Appreciate your understanding and continued support of YOLOv8.

**Comment by JDups** [comment]
Hello @glenn-jocher thank you for the response.

> Regarding the iterations parameter, it should indeed define the number of search iterations for hyperparameter tuning, rather than the number of training trials.

Reading the [YOLO documentation](https://docs.ultralytics.com/integrations/ray-tune/#usage) again it confirms that the iteration parameter sets the number of trials.
![image](https://github.com/ultralytics/ultralytics/assets/116138053/aa2a8a32-04f1-4ccb-8af0-871cea155071)

This matches the output produced when calling the tune method. And it also seems to match up with what the Ray Tune documentation describes as a trial. A trial would be the evaluation of a hyperparameter combination as they talk about [here](https://docs.ray.io/en/latest/tune/examples/tune_analyze_results.html).

![image](https://github.com/ultralytics/ultralytics/assets/116138053/5f0e9ad4-f2c1-46c1-986c-09af2a132da0)

After running the tuning the results are 10 trials of a single iteration each, as shown below. Which makes the plots like in their docs [here](https://docs.ray.io/en/latest/tune/examples/tune_analyze_results.html#plotting-metrics) not possible.

![image](https://github.com/ultralytics/ultralytics/assets/116138053/1a1e3f7a-eb71-43ea-9f9d-40dc89657cd4)

Also none of those ten trials terminated early eve though `grace_period` was set to 10. Many of them had really bad results and I would have expected the early stopping to activate. This combined with the already mentioned issues gives me the impression that the Ray Tune integration isn't functioning correctly.

**Comment by glenn-jocher** [comment]
Apologies for any confusion caused earlier, @JDups. You're right, the parameter in question does indeed set the number of trials rather than iterations. A trial typically corresponds to the evaluation of a specific hyperparameter combination.

If each trial in your `result_grid` only contains a single iteration and early stopping isn't activating as expected, it suggests that there might be an issue with the Ray Tune integration.

I recommend double-checking your tuning setup:
- Ensure your configuration is correct and that the model is training for the desired number of epochs per trial.
- Verify that the `grace_period` parameter is set up according to the Ray Tune documentation.
- Look into the status of trials and results in the `result_grid` to confirm whether the data logged per trial is as expected.

We will review the current Ray Tune integration to ensure it aligns with both the Ultralytics and Ray Tune documentation, making any necessary corrections.

Thank you for your patience and for bringing this to our attention. Your insights are invaluable in improving the YOLOv8 experience for everyone.

**Comment by Githubprojectsacc** [comment]
Hello! I want to get the metric values like map score for object detection for hyper parameter tuning using yolov8 model. I used the same code the above for finetuning with ASHA and hyperopt search algorithm. In the training function I returned the evaluation score .But in the result I got the same metric values in every trial .Can u pls help me with this.

**Comment by glenn-jocher** [comment]
@Githubprojectsacc hello!

Thanks for reaching out. 😊 It sounds like you're on the right track, but if you're seeing identical metric values across all trials, it could be a sign that the trials aren't varying your model's hyperparameters as intended, or the model state isn't being reset between trials.

Here's a quick suggestion: Ensure each trial uses a unique set of hyperparameters and that the model and optimizer states are reset or reinitialized before each trial begins. This helps to provide each trial a fresh start.

Here's a simplified example for defining your training function:

```python
from ultralytics import YOLO

def train_hyperopt(config):
    # Assume 'config' contains hyperparameters, e.g., 'lr': 0.001

    # Load model with specific hyperparameters
    model = YOLO('path/to/model.yaml', custom_hyperparams=config)

    # Your training logic here
    # ...

    # Return evaluation score, ensure it varies with 'config'
    return {"score": evaluation_score}
```

Make sure your configuration passed to `train_hyperopt(config)` changes for each trial. If using Ray Tune with ASHA or Hyperopt, the framework should handle this for you, as long as you've correctly configured the search space.

If the problem persists, double-check your evaluation logic and ensure the metric is correctly calculated and returned for each trial.

Keep experimenting, and feel free to share more code snippets if you continue facing issues. We're here to help! 🚀

**Comment by Githubprojectsacc** [comment]
Thank you!@glenn-jocher 
from ultralytics import YOLO
```
import os
import ray
from ray import tune
from ray import train
from ray.tune import ResultGrid
from ray.train import Result
from ray.tune.schedulers import ASHAScheduler
from ray.tune.search.hyperopt import HyperOptSearch

def model_train(config):
    model = YOLO("yolov8n.pt")
    data = '/content/drive/MyDrive/data.yaml'
    momentum=config["momentum"]
    batch=config["batch_size"]
    for epoch in range(1):
     trained_model = model.train(data=data,
                            epochs=1,
                            batch=batch,
                            workers=8,
                            imgsz=640,
                            patience=1,
                            momentum=momentum,
                          )
     eval=model.val()
     result=eval.results_dict
     metrics=eval.box.mp
     newmetric=momentum*2
     metricsnew=eval.box.map75
     train.report({"_metric":trained_model.box.mp})


ray.init(ignore_reinit_error=True)

search_space = {
    
    "batch_size": tune.choice([8, 16, 32]),
    "momentum": tune.loguniform(0.7, 0.98, 0.3)
    "batch_size": tune.choice([8, 16, 32]),
    "num_layers": tune.choice([3, 4, 5]),
    "activation": tune.choice(["relu", "tanh"]),
    "optimizer":tune.choice(['SGD', 'Adam', 'AdamW', 'NAdam', 'RAdam', 'RMSProp']),
}


search_alg = HyperOptSearch(metric="_metric")
num_samples=4

tuner = tune.Tuner(tune.with_resources(model_train,resources={"cpu":1,"gpu":1}),
    tune_config=tune.TuneConfig(
        metric="_metric",
        mode="min",
        search_alg=search_alg,
        num_samples=num_samples,
    ),
    param_space=search_space,
)
results = tuner.fit()

# Clean up Ray resources
ray.shutdown()
```
In this code I got the metrics value same for each trial the same precision score.I checked with other way of passing newmetric value which is the customized value that i given,whereas it shows different results.But there is some problem with the training method that I had implemented.The evaluation metrics are returning same for each trial.Can u pls guide me regarding this

**Comment by glenn-jocher** [comment]
Hey there! 🌟 Thanks for reaching out with your code snippet. It looks like you have a comprehensive setup for hyperparameter tuning with Ray Tune. If your metrics aren't varying across trials as expected, here are a few suggestions that might help:

1. **Verify Variable Updates:** Ensure that the hyperparameters `batch` and `momentum` passed from your config to the YOLO model's `train` method are actually influencing the training process. This might require adjusting how you initialize or pass these parameters to the `YOLO.train` method.

2. **Reset State:** Check if you need to reinitialize or reset the model state at the beginning of each `model_train` call. Persistent model states across trials can lead to the same results being reported.

3. **Review Evaluation:** Make sure the `model.val()` method is evaluated on a validation set that reflects the potential improvements or the lack thereof from the model training in each trial. 

4. **Typos and Duplicates**: You have a duplicate key `batch_size` in your `search_space` dictionary, and a missing comma after the `"momentum"` line.

5. **Logging the Correct Metric**: The line `train.report({"_metric":trained_model.box.mp})` seems to be reporting a metric directly from `trained_model` which might not be valid. You probably meant to report `metrics` or `metricsnew`.

Here's a quick fix to point 5:
```python
     train.report({"_metric": metrics})
```

And, don't forget to fix the syntax in your `search_space`:
```python
search_space = {
    "momentum": tune.loguniform(0.7, 0.98),
    "batch_size": tune.choice([8, 16, 32]),
    "num_layers": tune.choice([3, 4, 5]),
    "activation": tune.choice(["relu", "tanh"]),
    "optimizer":tune.choice(['SGD', 'Adam', 'AdamW', 'NAdam', 'RAdam', 'RMSProp']),
}
```

Give these tweaks a try and see if your metrics start showing the variation you expect. 😊 Keep experimenting and happy coding!

**Comment by proevgenii** [comment]
@Githubprojectsacc 
Did you solved and issue with identical resutls?
As soon as I'm using custom `search_space`, I'm getting completely identical metrics across all Trial runs
